datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/micrometer/DataSourceMetrics.java

```diff
@@ -0,0 +1,106 @@
+package com.github.gavlyukovskiy.micrometer;
+
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.stats.hist.Histogram;
+import io.micrometer.core.instrument.stats.quantile.WindowSketchQuantiles;
+import org.springframework.boot.autoconfigure.jdbc.metadata.DataSourcePoolMetadata;
+
+import javax.sql.DataSource;
+
+import java.sql.Connection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * {@link DataSource} metrics linked to micrometer's {@link MeterRegistry}.
+ *
+ * @author Arthur Gavlyukovskiy
+ * @since 1.2.2
+ */
+public class DataSourceMetrics {
+
+    private final AtomicInteger activeConnections = new AtomicInteger();
+    private final AtomicInteger totalConnections = new AtomicInteger();
+    private final AtomicInteger pendingConnections = new AtomicInteger();
+    private final Map<Connection, Long> connectionAcquireTimestamp = new ConcurrentHashMap<>();
+
+    private final Timer connectionObtainTimer;
+    private final DistributionSummary connectionUsageSummary;
+    private final DistributionSummary connectionCreatedSummary;
+    private final DistributionSummary connectionFailedSummary;
+
+    DataSourceMetrics(String dataSourceName, MeterRegistry meterRegistry, DataSourcePoolMetadata poolMetadata) {
+        connectionObtainTimer = Timer.builder("data.source.connection.wait")
+                .tags("pool", dataSourceName)
+                .register(meterRegistry);
+
+        connectionUsageSummary = DistributionSummary.builder("data.source.connection.usage")
+                .quantiles(WindowSketchQuantiles.quantiles(0.5, 0.95).create())
+                .histogram(Histogram.linearTime(TimeUnit.MILLISECONDS, 0, 10, 20))
+                .tags("pool", dataSourceName)
+                .register(meterRegistry);
+
+        connectionCreatedSummary = DistributionSummary.builder("data.source.connection.created")
+                .quantiles(WindowSketchQuantiles.quantiles(0.5, 0.95).create())
+                .histogram(Histogram.linearTime(TimeUnit.MILLISECONDS, 0, 10, 20))
```

Comment:
I'd recommend `Histogram.percentilesTime` for aggregable percentile approximations, as it will give you a reasonably tight error bound on almost all real world timers, including this one.